### PR TITLE
add new board --- esp32-s3-lcd-ev-board

### PR DESCRIPTION
**Pull request by Lzw655**
# Checklist for new Board Support package or Component

- [x] Component contains License
- [x] Component contains README.md
- [x] Project [README.md](../README.md) updated
- [x] Component contains idf_component.yml file with `url` field defined
- [x] Component was added to CI [upload job](https://github.com/espressif/esp-bsp/blob/master/.github/workflows/upload_component.yml#L17)
- [x] New files were added to CI build job
- [ ] _Optional:_ Component contains unit tests
- [x] CI passing

# Change description
Add a new BSP for ESP32-S3-LCD-Ev-Board, which supports below features:

- Support 800x480 RGB LCD (Default)
- Support 480x480 RGB LCD (Need to enable `BSP_LCD_SUB_BOARD_480_480` in menuconfig)
- Support to avoid tearing effect (Need to enable `BSP_LCD_RGB_DOUBLE_BUFFER`, `BSP_LCD_RGB_REFRESH_TASK_ENABLE` and `BSP_DISPLAY_LVGL_AVOID_TEAR` in menuconfig)

The BSP should be compiled with IDF release/v5.0 or later. The patch of [PSRAM_Octal_120M](https://github.com/Lzw655/esp-bsp/blob/test_esp32-s3-lcd-ev-board/examples/test_lvgl_demos/patch/release_v5.0_PSRAM_120MHz.patch)(only for release/v5.0) can be applied to let the RGB LCD achieve the best performance.

### CODE REVIEW

**Comment by Lzw655** [comment]
@espzav @tore-espressif This BSP should support 800x480 and 480x480 RGB LCD now. Because these LCDs are the most popular among customers. PTAL, thank you very much.

**Comment by espzav** [comment]
Hi @Lzw655, I started the review and I wanted to try it on the board first. I changed the `display_rotation` example for using your BSP. I have `ESP32-S3-LCD_Ev_Board_SUB2_V1.1` it seems, that it is 480x480 RGB LCD. I selected it and I selected Octal PSRAM. 
I have these issues:
1. Missing function `bsp_display_backlight_on()`.
2. These:
```
C:/WORK/GIT Temp/LZW655/esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:120:21: warning: initialization of '_Bool (*)(struct esp_lcd_panel_t *, const esp_lcd_rgb_panel_event_data_t *, void *)' from incompatible pointer type '_Bool (*)(struct esp_lcd_panel_t *, esp_lcd_rgb_panel_event_data_t *, void *)' [-Wincompatible-pointer-types]
  120 |         .on_vsync = on_vsync_event,
      |                     ^~~~~~~~~~~~~~
C:/WORK/GIT Temp/LZW655/esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:120:21: note: (near initialization for 'cbs.on_vsync')
C:/WORK/GIT Temp/LZW655/esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c: In function 'spi_write':
C:/WORK/GIT Temp/LZW655/esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:17:25: error: implicit declaration of function 'ets_delay_us'; did you mean 'esp_rom_delay_us'? [-Werror=implicit-function-declaration]
   17 | #define udelay(t)       ets_delay_us(t)
      |                         ^~~~~~~~~~~~
C:/WORK/GIT Temp/LZW655/esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:203:9: note: in expansion of macro 'udelay'
  203 |         udelay(10);
      |         ^~~~~~
C:/WORK/GIT Temp/LZW655/esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c: In function 'lcd_config':
C:/WORK/GIT Temp/LZW655/esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:371:1: error: control reaches end of non-void function [-Werror=return-type]
  371 | }
```
Please, could you check it? Or help me, if I doing something wrong?
Thank you.

**Comment by Lzw655** [comment]
> Hi @Lzw655, I started the review and I wanted to try it on the board first. I changed the `display_rotation` example for using your BSP. I have `ESP32-S3-LCD_Ev_Board_SUB2_V1.1` it seems, that it is 480x480 RGB LCD. I selected it and I selected Octal PSRAM. I have these issues:
> 
> 1. Missing function `bsp_display_backlight_on()`.
> 2. These:
> 
> ```
> C:/WORK/GIT Temp/LZW655/esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:120:21: warning: initialization of '_Bool (*)(struct esp_lcd_panel_t *, const esp_lcd_rgb_panel_event_data_t *, void *)' from incompatible pointer type '_Bool (*)(struct esp_lcd_panel_t *, esp_lcd_rgb_panel_event_data_t *, void *)' [-Wincompatible-pointer-types]
>   120 |         .on_vsync = on_vsync_event,
>       |                     ^~~~~~~~~~~~~~
> C:/WORK/GIT Temp/LZW655/esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:120:21: note: (near initialization for 'cbs.on_vsync')
> C:/WORK/GIT Temp/LZW655/esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c: In function 'spi_write':
> C:/WORK/GIT Temp/LZW655/esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:17:25: error: implicit declaration of function 'ets_delay_us'; did you mean 'esp_rom_delay_us'? [-Werror=implicit-function-declaration]
>    17 | #define udelay(t)       ets_delay_us(t)
>       |                         ^~~~~~~~~~~~
> C:/WORK/GIT Temp/LZW655/esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:203:9: note: in expansion of macro 'udelay'
>   203 |         udelay(10);
>       |         ^~~~~~
> C:/WORK/GIT Temp/LZW655/esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c: In function 'lcd_config':
> C:/WORK/GIT Temp/LZW655/esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:371:1: error: control reaches end of non-void function [-Werror=return-type]
>   371 | }
> ```
> 
> Please, could you check it? Or help me, if I doing something wrong? Thank you.

Hi @espzav, thank you to start the review.
1. This board's hardware doesn't support to control brightness of LCD. But I add related functions in the new [commit](https://github.com/espressif/esp-bsp/pull/89/commits/62c7a3afa6283a5df33a1a72ffb07a5935cabd5d) for compatibility.
2. Maybe it is caused by the different IDF versions. I modify some code in the new [commit](https://github.com/espressif/esp-bsp/pull/89/commits/ac8803a6ca4299cc343df816a01022f9759cbb1f) to avoid these warnings and errors. And it can work well in IDF release/v5.0.

**Comment by Lzw655** [comment]
I add two examples to test this BSP([test_display_rotation](https://github.com/Lzw655/esp-bsp/tree/test_esp32-s3-lcd-ev-board/examples/test_display_rotation) and [test_lvgl_demos](https://github.com/Lzw655/esp-bsp/tree/test_esp32-s3-lcd-ev-board/examples/test_lvgl_demos)). You can directly compile and upload to the board. Please make sure the IDF version is release/v5.0 or later.

**Comment by Lzw655** [comment]
> @Lzw655 Thank you for changes and fixes. Now it is working. I will be continue with code review now.

Ok, thanks a lot.

**Review by espzav — COMMENTED** [review]
Nice work, thank you! Sometimes, it was hard to understand for me, it is really complex. I still haven't got lot of experiences with RGB displays. I think, someone better should review this too (maybe @suda-morris can help). 
I left some comments.

File: esp32_s3_lcd_ev_board/README.md
Comment (Lzw655) [reply esp32_s3_lcd_ev_board/README.md]: You're right. It's better to add a table to show the details about the LCD sub-board. Thanks for your suggestions. I add two tables [here](https://github.com/espressif/esp-bsp/pull/89/commits/278946e1027c83770a9746e38f279566c51418e5#diff-1a13ae2d8f3402d40ef16fb3add1e94f61f5b0dacd362426c1dff32dfc3c1bf4R220-R233) and [here](https://github.com/espressif/esp-bsp/pull/89/commits/278946e1027c83770a9746e38f279566c51418e5#diff-8e8bed615b97420f29dc0e53bb4779337a675f191f128b9e85f46fed8f708c8dR15-R21). Please check it out if there is anything need to add.

File: esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c
Comment (Lzw655) [reply esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:358]: Yeah, it also confused me at first. This LCD type is called "3 Wire SPI + Parallel RGB Interface". You can read this [article](https://focuslcds.com/3-wire-spi-parallel-rgb-interface-fan4213/) to understand it.

File: esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c
Comment (Lzw655) [reply esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:7]: Fixed [here](https://github.com/espressif/esp-bsp/pull/89/commits/278946e1027c83770a9746e38f279566c51418e5#diff-d6e640338b4dc95a1d9e90b81b17d1ac1a0d001f50977d8a52be5e2895838044R1-R6).

File: esp32_s3_lcd_ev_board/sub_board/sub_board_800x480.c
Comment (Lzw655) [reply esp32_s3_lcd_ev_board/sub_board/sub_board_800x480.c:7]: Fixed [here](https://github.com/espressif/esp-bsp/pull/89/commits/278946e1027c83770a9746e38f279566c51418e5#diff-43098840b47cfd6e4dfb4db2d0ff0af2f2ba8569a38a096ebc253347cc4e328eR1-R6).

**Comment by Lzw655** [comment]
@suda-morris Please review this PR. Thank you very much.

File: esp32_s3_lcd_ev_board/README.md
Comment (espzav) [reply esp32_s3_lcd_ev_board/README.md]: Great!

File: esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c
Comment (espzav) [reply esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:358]: I am thinking about this controller (GC9503CV) and is it possible make it as separate esp_lcd component? It could be, if the SPI lines wouldn't on IO expander. But I don't know, if we can do it in this case... What do you think?

Comment (Lzw655) [reply esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:358]: Yeah, we can make it as separate esp_lcd component. But as you say, the SPI lines are fixed on IO expander. Actually, this board is limited by a lack of IO as it needs to have I2S, I2C, USB, and Octal SPSRAM at the same time. Fortunately, the SPI function of the LCD is only used for register configuration, and the transmission rate is not high, so it can be simulated by the pin of the IO expander.

Comment (Lzw655) [reply esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:358]: What's more, the `esp_lcd` driver in IDF doesn't support 3-wire (also called 9-bit) SPI now. You can take a look [here](https://github.com/espressif/esp-idf/blob/release/v5.0/components/esp_lcd/src/esp_lcd_panel_io_spi.c#L69-L71).

Comment (espzav) [reply esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:358]: Ok, I understand. Thank you for info.

**Comment by Lzw655** [comment]
@tore-espressif PTAL, thank you very much!

**Comment by espzav** [comment]
One additional think: We should add this board into main readme file. @tore-espressif what do you think?

File: esp32_s3_lcd_ev_board/priv_include/bsp_sub_board.h
Comment (tore-espressif) [inline esp32_s3_lcd_ev_board/priv_include/bsp_sub_board.h:12]: sdkconfig.h is not needed in this file

File: esp32_s3_lcd_ev_board/CMakeLists.txt
Comment (tore-espressif) [inline esp32_s3_lcd_ev_board/CMakeLists.txt:11]: Could you please add `PRIV_INCLUDE_DIRS priv_include` ? That way we can get rid of the ugly `#include "../priv_include/bsp_err_check.h"`  includes in `sub_board` directory

File: esp32_s3_lcd_ev_board/sub_board/sub_board_800x480.c
Comment (tore-espressif) [inline esp32_s3_lcd_ev_board/sub_board/sub_board_800x480.c:18]: Please define the `priv_include` directory as private include in CMakeLists.txt so we can get rid of the relative `..` in path

File: esp32_s3_lcd_ev_board/idf_component.yml
Comment (tore-espressif) [inline esp32_s3_lcd_ev_board/idf_component.yml:22]: Isn't ADC ES7210 missing here?

File: esp32_s3_lcd_ev_board/include/bsp/esp32_s3_lcd_ev_board.h
Comment (tore-espressif) [inline esp32_s3_lcd_ev_board/include/bsp/esp32_s3_lcd_ev_board.h:126]: Hm, usually we don't provide wrapper functions around dependencies of BSP. The users can init the expander them selves by calling `esp_io_expander_new_i2c_tca9554`.

The idea is that if we provide only the `init` wrapper, the API look incomplete (the rest of the API is defined in tca9554 component).

What do you think?

_EDIT:_ Now I see that you need to use the expander to enable the audio amplifier... who should own the expander handle then?

File: esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c
Comment (tore-espressif) [reply esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:358]: Is this resolved @espzav ?

**Review by tore-espressif — CHANGES_REQUESTED** [review]
@Lzw655 I'm sorry for the late review. I left few comments in the code, that I think should be addressed.

There are other two points:

- Please update the main README https://github.com/espressif/esp-bsp/blob/master/README.md
- Currently the new files are not compiled in CI job. (BSPs are compiled only with examples, and this BSP does not have any example yet). But since we want to include this BSP in upcoming SquareLine release, we can do it in a follow-up MR. What do you think @espzav ?

**Comment by espzav** [comment]
@Lzw655 I am preparing SquareLine files. Please, is anywhere images of the both supported screens? I need it for SquareLine projects and I cannot found it.

File: esp32_s3_lcd_ev_board/Kconfig
Comment (espzav) [inline esp32_s3_lcd_ev_board/Kconfig:110]: I tried this BSP with SquareLine - used their Thermostat example and it was really slow. I had to changed this to 100, for faster. I think, it can be more. Why the default is 10? Could it be defoult more (e.g. 100) ?

**Comment by Lzw655** [comment]
> @Lzw655 I'm sorry for the late review. I left few comments in the code, that I think should be addressed.
> 
> There are other two points:
> 
> * Please update the main README https://github.com/espressif/esp-bsp/blob/master/README.md
> * Currently the new files are not compiled in CI job. (BSPs are compiled only with examples, and this BSP does not have any example yet). But since we want to include this BSP in upcoming SquareLine release, we can do it in a follow-up MR. What do you think @espzav ?

@tore-espressif  Thank you for pointing out mistakes and giving so many suggestions。

* It's been updated, PTLK [here](https://github.com/espressif/esp-bsp/pull/89/commits/abc7099df435b938ce2cd02f1ecafc7ca546c79c#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R15-R48).
* I add an example named "display_lvgl_demos", PTAL.

File: esp32_s3_lcd_ev_board/priv_include/bsp_sub_board.h
Comment (Lzw655) [reply esp32_s3_lcd_ev_board/priv_include/bsp_sub_board.h]: Fixed, PTAL [here](https://github.com/espressif/esp-bsp/pull/89/commits/abc7099df435b938ce2cd02f1ecafc7ca546c79c#diff-825e2d50928be7919de09e2fcbdf9f5d8967d785e1e7134d9e9bde04fc8ea5d6R9-R11).

File: esp32_s3_lcd_ev_board/CMakeLists.txt
Comment (Lzw655) [reply esp32_s3_lcd_ev_board/CMakeLists.txt:11]: Fixed, PTAL [here](https://github.com/espressif/esp-bsp/pull/89/commits/abc7099df435b938ce2cd02f1ecafc7ca546c79c#diff-f87fc8945850544b13520a5fb8f70dde17fd0ef30f1507472166d6659d5b5e98R12).

File: esp32_s3_lcd_ev_board/sub_board/sub_board_800x480.c
Comment (Lzw655) [reply esp32_s3_lcd_ev_board/sub_board/sub_board_800x480.c]: Fixed, PTAL [1](https://github.com/espressif/esp-bsp/pull/89/commits/8723853de84500501a915f9a2c568d9e528a8b67#diff-43098840b47cfd6e4dfb4db2d0ff0af2f2ba8569a38a096ebc253347cc4e328eL17-R18), [2](https://github.com/espressif/esp-bsp/pull/89/commits/abc7099df435b938ce2cd02f1ecafc7ca546c79c#diff-d6e640338b4dc95a1d9e90b81b17d1ac1a0d001f50977d8a52be5e2895838044R19-R20) and [3](https://github.com/espressif/esp-bsp/pull/89/commits/abc7099df435b938ce2cd02f1ecafc7ca546c79c#diff-521a0f3ed3e2e3565f9da746eb3de4eed006957460f7cd6077b1491d48d90f60R23-R24).

File: esp32_s3_lcd_ev_board/idf_component.yml
Comment (Lzw655) [reply esp32_s3_lcd_ev_board/idf_component.yml:22]: Added, PTAL [here](https://github.com/espressif/esp-bsp/pull/89/commits/abc7099df435b938ce2cd02f1ecafc7ca546c79c#diff-293e140e94a241f4e6688d73438f3c05d3ce7221850cfef65ec351dc7f4fd080R23-R25).

File: esp32_s3_lcd_ev_board/include/bsp/esp32_s3_lcd_ev_board.h
Comment (Lzw655) [reply esp32_s3_lcd_ev_board/include/bsp/esp32_s3_lcd_ev_board.h:126]: And if sub_board_480x480 is used, it's necessary to init tca9554, which you can see [here](https://github.com/espressif/esp-bsp/pull/89/commits/abc7099df435b938ce2cd02f1ecafc7ca546c79c#diff-521a0f3ed3e2e3565f9da746eb3de4eed006957460f7cd6077b1491d48d90f60R278-R280). So I left API `bsp_io_expander_init` for users to init tca9554 or just get the handle of it.

File: esp32_s3_lcd_ev_board/Kconfig
Comment (Lzw655) [reply esp32_s3_lcd_ev_board/Kconfig]: I didn't think too much and just use the minimum size of the LVGL buffer. Sure, I can increase it to 100, PTAL [here](https://github.com/espressif/esp-bsp/pull/89/commits/927713d146b33c808aa07e634235995d44adb973#diff-918c0433b106ae9d861b42e18180e71a6a564c01719cbe2bcb38f2984a8dcf17L110-R110).

File: esp32_s3_lcd_ev_board/priv_include/bsp_sub_board.h
Comment (Lzw655) [reply esp32_s3_lcd_ev_board/priv_include/bsp_sub_board.h]: > @Lzw655 I am preparing SquareLine files. Please, are anywhere images of the both supported screens? I need it for SquareLine projects and I cannot find it.

Our colleague Zhang Wei has created a squareline project for this development board, how about asking him to raise a PR?

File: esp32_s3_lcd_ev_board/Kconfig
Comment (espzav) [reply esp32_s3_lcd_ev_board/Kconfig]: Thank you.

File: esp32_s3_lcd_ev_board/priv_include/bsp_sub_board.h
Comment (espzav) [reply esp32_s3_lcd_ev_board/priv_include/bsp_sub_board.h]: > Our colleague Zhang Wei has created a squareline project for this development board, how about asking him to raise a PR?
@Lzw655 Oh, I didn't know, that somebody prepared it. It is a little hurry, because, there is waiting customer for it and the SquareLine relese will be soon. I made it yesterday here: https://github.com/espressif/esp-bsp/tree/SquareLine/add_esp32_s3_ev_board/SquareLine/boards but we can make PR on work of your colleague. It douesn't matter, which branch will be merged, the result should be same :-) I was only waiting for merge this BSP...

File: esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c
Comment (espzav) [inline esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:361]: Only note for future. I was talking about IO expanders and LCD with @igrr and he bring two ideas:
1. Add user callbacks into esp_lcd and call it for all GPIOs (RST, CS, etc...)
2. Extend GPIO library for calling bigger pin numbers, which will be mapped as IO expanded by I2C
@tore-espressif We should select one choice in future and make it.

**Review by espzav — APPROVED** [review]
I checked it. Tested on HW the LVGL demos example. Only left one note for future. @Lzw655 Thank you. LGTM

File: esp32_s3_lcd_ev_board/include/bsp/esp32_s3_lcd_ev_board.h
Comment (tore-espressif) [reply esp32_s3_lcd_ev_board/include/bsp/esp32_s3_lcd_ev_board.h:126]: OK, thank you

File: esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c
Comment (tore-espressif) [reply esp32_s3_lcd_ev_board/sub_board/sub_board_480x480.c:361]: Without thinking too much about it, no 2 sounds better to me. But we need to discuss it with 'chip-support' team

**Review by tore-espressif — APPROVED** [review]

**Comment by tore-espressif** [comment]
@espzav @Lzw655 Thank you very much for your contributions and testing.

You can merge this, if you don't plan any more changes

**Pull request merged**
